// @nomiclabs/hardhat-vyper/dist/src/task-names.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TASK_COMPILE_VYPER_LOG_COMPILATION_RESULT = exports.TASK_COMPILE_VYPER_LOG_DOWNLOAD_COMPILER_END = exports.TASK_COMPILE_VYPER_LOG_DOWNLOAD_COMPILER_START = exports.TASK_COMPILE_VYPER_GET_SOURCE_PATHS = exports.TASK_COMPILE_VYPER_GET_SOURCE_NAMES = exports.TASK_COMPILE_VYPER_READ_FILE = exports.TASK_COMPILE_VYPER_GET_BUILD = exports.TASK_COMPILE_VYPER_RUN_BINARY = exports.TASK_COMPILE_VYPER = void 0;
exports.TASK_COMPILE_VYPER = "compile:vyper";
exports.TASK_COMPILE_VYPER_RUN_BINARY = "compile:vyper:binary:run";
exports.TASK_COMPILE_VYPER_GET_BUILD = "compile:vyper:binary:get-build";
exports.TASK_COMPILE_VYPER_READ_FILE = "compile:vyper:read-file";
exports.TASK_COMPILE_VYPER_GET_SOURCE_NAMES = "compile:vyper:get-source-names";
exports.TASK_COMPILE_VYPER_GET_SOURCE_PATHS = "compile:vyper:get-source-paths";
exports.TASK_COMPILE_VYPER_LOG_DOWNLOAD_COMPILER_START = "compile:vyper:log:download-compiler-start";
exports.TASK_COMPILE_VYPER_LOG_DOWNLOAD_COMPILER_END = "compile:vyper:log:download-compiler-end";
exports.TASK_COMPILE_VYPER_LOG_COMPILATION_RESULT = "compile:vyper:log:compilation-result";
